core/unit/unit.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package unit
+
+import "strconv"
+
+// UnitState contains the various state saved for this unit,
+// including from the charm itself and the uniter.
+type UnitState struct {
+	// State encodes the unit's persisted state as a list of key-value pairs.
+	State map[string]string
+
+	// UniterState is a serialized yaml string containing the uniters internal
+	// state for this unit.
+	UniterState string
+
+	// RelationState is a serialized yaml string containing relation internal
+	// state for this unit from the uniter.
+	RelationState map[int]string
+
+	// StorageState is a serialized yaml string containing storage internal
+	// state for this unit from the uniter.
+	StorageState string
+}
+
+// BSONRelationData makes the RelationState data BSON friendly by
+// translating the int map key to a string key.
+func (u UnitState) BSONRelationData() map[string]string {
```

Comment:
I think this should be an internal implementation detail of the serialization logic in the state package similar to how we escape map keys to make things mongo-friendly.

Reply:
moved